reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,486 @@
+import com.vaticle.typedb.core.pattern.constraint.thing.RelationConstraint;
+import com.vaticle.typedb.core.pattern.variable.ThingVariable;
+import com.vaticle.typedb.core.pattern.variable.TypeVariable;
+import com.vaticle.typedb.core.pattern.variable.Variable;
+import com.vaticle.typedb.core.traversal.common.Identifier;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.vaticle.typedb.common.collection.Collections.list;
+import static com.vaticle.typedb.common.collection.Collections.set;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public class AnswerCountEstimator {
+    private final LogicManager logicMgr;
+    private final Map<ResolvableConjunction, ConjunctionAnswerCountEstimator> estimators;
+
+    // Cycle handling
+    private final ArrayList<ResolvableConjunction> initializationStack;
+    private final Set<ResolvableConjunction> cycleHeads;
+    private final Set<ResolvableConjunction> toReset;
+    private final AnswerCountModel answerCountModel;
+
+    public AnswerCountEstimator(LogicManager logicMgr) {
+        this.logicMgr = logicMgr;
+        this.answerCountModel = new AnswerCountModel(this);
+        this.estimators = new HashMap<>();
+        this.initializationStack = new ArrayList<>();
+        this.cycleHeads = new HashSet<>();
+        this.toReset = new HashSet<>();
+    }
+
+    public long estimateAllAnswers(ResolvableConjunction conjunction) {
+        registerConjunctionAndInitialize(conjunction);
+        return estimators.get(conjunction).estimateAllAnswers();
+    }
+
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variableFilter) {
+        return estimateAnswers(conjunction, variableFilter, logicMgr.compile(conjunction));
+    }
+
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variableFilter, Set<Resolvable<?>> includedResolvables) {
+        registerConjunctionAndInitialize(conjunction);
+        return estimators.get(conjunction).estimateAnswers(variableFilter, includedResolvables);
+    }
+
+    private void registerConjunctionAndInitialize(ResolvableConjunction conjunction) {
+        if (!estimators.containsKey(conjunction)) {
+            estimators.put(conjunction, new ConjunctionAnswerCountEstimator(this, conjunction));
+        }
+
+        if (estimators.get(conjunction).mayInitialize()) {
+            initializationStack.add(conjunction);
+            estimators.get(conjunction).initialize();
+            assert initializationStack.get(initializationStack.size() - 1) == conjunction;
+            initializationStack.remove(initializationStack.size() - 1);
```

Comment:
correct me if I'm wrong, but the fact that the `stack` is a class variable not a local variable that is passed through the recursion explicitly is for convenience?

What keeping it as a class variable is that it looks like somehow the stack is going to be important in the next invocation by the user of `estimateAnswers`... however, here we clearly `add` and then `remove` the same item from the stack!

I think the right process would be to pass the stack into the `registerConjunctionAndInitialise` method. The entry points (`estimateAllAnswers` etc.) will initialise an empty stack to use.

Reply:
Then the AnswerCountEstimators could be aware of their place in the graph, which they don't need to be.
Is it worth passing an argument to a class if it's only going to be passed back to this one?

Reply:
I think it makes sense:
```computeAnswers() {
  visited = new Stack()
  estimator.computeEstimate(visited)
}

Estimator {

  computeEstimate(visited) {
    visited.add(conjunction)
    computeRecursiveEstimates(visited)
    visited.pop()
    }
 }
```

Reply:
That does look nice and pure.
But now, Should I also pass cycle_heads and reset through as well? Because they're all related to each other. I could also wrap them into a context object.

Reply:
do `cycle_heads` and `reset` affect the operation of the next call to `computeAnswers`? I believe so right? In that case, they're still ok. The `stack` however is purely used within a single recursive computation.

Will have to see how it looks after this change

Reply:
They all work together in a single-call to do cycle-detection. By the time the stack is empty, reset and cycle_heads will be empty.